### PR TITLE
Add support for client_certificate_v2 posture rule

### PR DESCRIPTION
**Pull request by jlu-cloudflare**
depends on https://github.com/cloudflare/cloudflare-go/pull/1685

This PR is a duplicate of: https://github.com/cloudflare/terraform-provider-cloudflare/pull/3231
This was reopened to create the PR off a feature branch instead of master.

### CODE REVIEW

File: internal/sdkv2provider/schema_cloudflare_device_posture_rule.go [hunk 1]
```diff
@@ -251,6 +251,43 @@ func resourceCloudflareDevicePostureRuleSchema() map[string]*schema.Schema {
...
+					},
+					"extended_key_usage": {
+						Type: schema.TypeSet,
+						Elem: &schema.Schema{
+							Type: schema.TypeString,
+						},
+						Description: fmt.Sprintf("List of values indicating purposes for which the certificate public key can be used. %s", renderAvailableDocumentationValuesStringSlice([]string{"clientAuth", "emailProtection"})),
+					},
+					"locations": {
+						Type: schema.TypeMap,
```
Comment (jacobbednarz) [inline internal/sdkv2provider/schema_cloudflare_device_posture_rule.go:267]: this is causing an issue

```
    resource_cloudflare_device_posture_rule_test.go:196: Step 1/1 error: Error running pre-apply plan: exit status 1
        
        Error: Invalid provider configuration
        
        Provider "registry.terraform.io/hashicorp/cloudflare" requires explicit
        configuration. Add a provider block to the root module and configure the
        provider's required arguments as described in the provider documentation.
        
        
        Error: InternalValidate
        
          with provider["registry.terraform.io/hashicorp/cloudflare"],
          on <empty> line 0:
          (source code not available)
        
        Internal validation of the provider failed! This is always a bug
        with the provider itself, and not a user issue. Please report
        this bug:
        
        resource cloudflare_device_posture_rule: locations: TypeMap with Elem
        *Resource not supported,use TypeList/TypeSet with Elem *Resource or TypeMap
        with Elem *Schema
```

File: internal/sdkv2provider/schema_cloudflare_device_posture_rule.go [hunk 2]
```diff
@@ -251,6 +251,43 @@ func resourceCloudflareDevicePostureRuleSchema() map[string]*schema.Schema {
 						Optional:    true,
 						Description: "The total score from Tanium.",
 					},
+					"check_private_key": {
+						Type:        schema.TypeBool,
+						Optional:    true,
+						Description: "Confirm the certificate was not imported from another device.",
+					},
+					"extended_key_usage": {
```
Comment (jacobbednarz) [inline internal/sdkv2provider/schema_cloudflare_device_posture_rule.go:259]: this is missing `Optional`, `Computed` or `Required`.

File: internal/sdkv2provider/resource_cloudflare_device_posture_rule.go [hunk 3]
```diff
@@ -264,6 +264,18 @@ func setDevicePostureRuleInput(rule *cloudflare.DevicePostureRule, d *schema.Res
 		if totalScore, ok := d.GetOk("input.0.total_score"); ok {
 			input.TotalScore = totalScore.(int)
 		}
+		if checkPrivateKey, ok := d.GetOk("input.0.check_private_key"); ok {
+			input.CheckPrivateKey = checkPrivateKey.(bool)
```
Comment (jacobbednarz) [inline internal/sdkv2provider/resource_cloudflare_device_posture_rule.go:268]:
```suggestion
			input.CheckPrivateKey = cloudflare.BoolPtr(checkPrivateKey.(bool))
```

**Comment by jacobbednarz** [comment]
we're hitting a bad value at https://github.com/cloudflare/terraform-provider-cloudflare/pull/3512/files#diff-10fbc0331ed0beacdaac28b15f0560b0804cf96ba3a41ea86f12bf271b80dbcdR102

```
TF_ACC=1 go test ./internal/sdkv2provider -v -run "^TestAccCloudflareDevicePostureRule_" -count 1 -timeout 120m -parallel 1
=== RUN   TestAccCloudflareDevicePostureRule_SerialNumber
panic: input.0.locations: '': source data must be an array or slice, got struct

goroutine 139 [running]:
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*ResourceData).Set(0x14000cffe00, {0x1018371e9, 0x5}, {0x101bf9e20, 0x14000890d20})
	/Users/jacob/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.34.0/helper/schema/resource_data.go:233 +0x24c
github.com/cloudflare/terraform-provider-cloudflare/internal/sdkv2provider.resourceCloudflareDevicePostureRuleRead({0x101ea2a70, 0x1400060acb0}, 0x14000cffe00, {0x101e82380?, 0x1400086a420})
	/Users/jacob/src/github.com/cloudflare/terraform-provider-cloudflare/internal/sdkv2provider/resource_cloudflare_device_posture_rule.go:102 +0x6e0
github.com/cloudflare/terraform-provider-cloudflare/internal/sdkv2provider.resourceCloudflareDevicePostureRuleCreate({0x101ea2a70, 0x1400060acb0}, 0x14000cffe00, {0x101e82380, 0x1400086a420})
	/Users/jacob/src/github.com/cloudflare/terraform-provider-cloudflare/internal/sdkv2provider/resource_cloudflare_device_posture_rule.go:78 +0x4e8
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).create(0x14000759c00, {0x101ea29c8, 0x14000667650}, 0x14000cffe00, {0x101e82380, 0x1400086a420})
	/Users/jacob/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.34.0/helper/schema/resource.go:806 +0xe4
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0x14000759c00, {0x101ea29c8, 0x14000667650}, 0x14000c56270, 0x14000cffc80, {0x101e82380, 0x1400086a420})
	/Users/jacob/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.34.0/helper/schema/resource.go:937 +0x884
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0x1400000e420, {0x101ea29c8?, 0x140006675c0?}, 0x140006cc640)
	/Users/jacob/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.34.0/helper/schema/grpc_provider.go:1153 +0xaa4
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ApplyResourceChange(0x140008befa0, {0x101ea29c8?, 0x1400063b6b0?}, 0x1400060a540)
	/Users/jacob/go/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.23.0/tfprotov5/tf5server/server.go:865 +0x2b4
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler({0x101e46ca0, 0x140008befa0}, {0x101ea29c8, 0x1400063b6b0}, 0x14000cff380, 0x0)
	/Users/jacob/go/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.23.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:518 +0x1c0
google.golang.org/grpc.(*Server).processUnaryRPC(0x14000a04800, {0x101ea29c8, 0x1400063b620}, {0x101ea8900, 0x14000b36480}, 0x14000af99e0, 0x14000e7afc0, 0x102a0b618, 0x0)
	/Users/jacob/go/pkg/mod/google.golang.org/grpc@v1.63.2/server.go:1369 +0xb58
google.golang.org/grpc.(*Server).handleStream(0x14000a04800, {0x101ea8900, 0x14000b36480}, 0x14000af99e0)
	/Users/jacob/go/pkg/mod/google.golang.org/grpc@v1.63.2/server.go:1780 +0xb20
google.golang.org/grpc.(*Server).serveStreams.func2.1()
	/Users/jacob/go/pkg/mod/google.golang.org/grpc@v1.63.2/server.go:1019 +0x8c
created by google.golang.org/grpc.(*Server).serveStreams.func2 in goroutine 166
	/Users/jacob/go/pkg/mod/google.golang.org/grpc@v1.63.2/server.go:1030 +0x13c
FAIL	github.com/cloudflare/terraform-provider-cloudflare/internal/sdkv2provider	2.959s
FAIL
make: *** [testacc] Error 1
```

if you run the same acceptance test, this _should_ pop up too.

**Comment by jlu-cloudflare** [comment]
Seems like the error is caused by running `(d *ResourceData) Set` and passing in a value with a nested struct (input.0.locations). 
I'm having trouble working around this issue as it doesn't seem like there's any way to convert the struct into a slice/array. Do you have any advice as to how I could resolve this?

**Comment by jlu-cloudflare** [comment]
I pushed a commit that should have addressed that error

**Comment by jacobbednarz** [comment]
looking good! thanks.

```
TF_ACC=1 go test ./internal/sdkv2provider -v -run "^TestAccCloudflareDevicePostureRule_" -count 1 -timeout 120m -parallel 1
=== RUN   TestAccCloudflareDevicePostureRule_SerialNumber
--- PASS: TestAccCloudflareDevicePostureRule_SerialNumber (5.00s)
=== RUN   TestAccCloudflareDevicePostureRule_OsVersion
--- PASS: TestAccCloudflareDevicePostureRule_OsVersion (3.80s)
=== RUN   TestAccCloudflareDevicePostureRule_OsVersionExtra
--- PASS: TestAccCloudflareDevicePostureRule_OsVersionExtra (3.87s)
=== RUN   TestAccCloudflareDevicePostureRule_LinuxOsDistro
--- PASS: TestAccCloudflareDevicePostureRule_LinuxOsDistro (4.70s)
=== RUN   TestAccCloudflareDevicePostureRule_DomainJoined
--- PASS: TestAccCloudflareDevicePostureRule_DomainJoined (4.39s)
=== RUN   TestAccCloudflareDevicePostureRule_Firewall
--- PASS: TestAccCloudflareDevicePostureRule_Firewall (3.80s)
=== RUN   TestAccCloudflareDevicePostureRule_DiskEncryption_RequireAll
--- PASS: TestAccCloudflareDevicePostureRule_DiskEncryption_RequireAll (4.19s)
=== RUN   TestAccCloudflareDevicePostureRule_DiskEncryption_CheckDisks
--- PASS: TestAccCloudflareDevicePostureRule_DiskEncryption_CheckDisks (4.21s)
=== RUN   TestAccCloudflareDevicePostureRule_Intune
    provider_test.go:217: Skipping acceptance test for default account (f037e56e89293a057740de681ac9abbe). Assertion requires active Intune license.
--- SKIP: TestAccCloudflareDevicePostureRule_Intune (0.00s)
PASS
ok  	github.com/cloudflare/terraform-provider-cloudflare/internal/sdkv2provider	35.116s
```

**Pull request merged**